### PR TITLE
add standard error library

### PR DESCRIPTION
**Pull request by imtbkcat**

### CODE REVIEW

File: terror_error.go [hunk 1]
```diff
@@ -122,16 +112,22 @@ func (e *Error) MessageTemplate() string {
 	return e.message
 }
 
+// SetErrCodeText sets the text error code for standard error.
+func (e *Error) SetErrCodeText(codeText string) *Error {
```
Comment (lonng) [inline terror_error.go:116]: Seems this function should be deleted.

**Review by lonng — COMMENTED** [review]
Rest LGTM

File: terror_error.go [hunk 2]
```diff
@@ -273,21 +260,83 @@ func (e *Error) UnmarshalJSON(data []byte) error {
...
-		Description: className,
-		ID:          err.Class,
-		registry:    &Registry{Name: regName},
-	}
 	return nil
 }
+
+func (e *Error) Wrap(err error) *Error {
+	newErr := *e
+	newErr.cause = err
```
Comment (amyangfei) [inline terror_error.go:276]: What should we do if  the wrapped `err` is nil

File: terror_error.go [hunk 3]
```diff
@@ -273,21 +260,83 @@ func (e *Error) UnmarshalJSON(data []byte) error {
...
+	root := Unwrap(e.cause)
+	if root == nil {
+		return e.cause
+	}
+	return root
+}
+
+func (e *Error) FastGenWithCause(args ...interface{}) error {
+	err := *e
+	err.message = e.cause.Error()
```
Comment (amyangfei) [inline terror_error.go:290]: will panic here if `e.cause` is nil

File: terror_error.go [hunk 4]
```diff
@@ -273,21 +260,83 @@ func (e *Error) UnmarshalJSON(data []byte) error {
...
+	err.args = args
+	return SuspendStack(&err)
+}
+
+func (e *Error) GenWithStackByCause(args ...interface{}) error {
+	err := *e
+	err.message = e.cause.Error()
```
Comment (amyangfei) [inline terror_error.go:297]: ditto

**Comment by amyangfei** [comment]
LGTM

**Review by lonng — APPROVED** [review]

**Review by lysu — APPROVED** [review]

File: terror_error.go [hunk 5]
```diff
@@ -128,10 +118,10 @@ func (e *Error) Error() string {
 	if len(describe) == 0 {
 		describe = ErrCodeText(strconv.Itoa(int(e.code)))
 	}
-	return fmt.Sprintf("[%s] %s", e.RFCCode(), e.getMsg())
+	return fmt.Sprintf("[%s]%s", e.RFCCode(), e.GetMsg())
```
Comment (lysu) [inline terror_error.go:121]: how about check `e == nil` in the first line

I ob a stranger question....

we often log some nil terror like this:

image

and fmt will meet panic and self recovery...

image

maybe better check nil and return "nil" string?

**Comment by lysu** [comment]
LGTM

**Pull request merged**